File: translated/f-to-c/DRB028-privatemissing-orig-yes.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//tmp should be annotated as private to avoid race condition.
//Data race pairs: tmp@28:9:W vs. tmp@29:16:R
//                 tmp@28:9:W vs. tmp@28:9:W

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

int main()
{
    int i, tmp, len;
    int* a;

    len = 100;
    a = (int*)malloc(len * sizeof(int));

    for (i = 0; i < len; i++) {
        a[i] = i + 1;
    }

    #pragma omp parallel for
    for (i = 0; i < len; i++) {
        tmp = a[i] + i + 1;  // Adjust for 0-based indexing
        a[i] = tmp;
    }

    printf("a(50) = %d\n", a[49]);

    free(a);

    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB028-privatemissing-orig-yes.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//tmp should be annotated as private to avoid race condition.
//Data race pairs: tmp@28:9:W vs. tmp@29:16:R
//                 tmp@28:9:W vs. tmp@28:9:W

package main

import (
	"fmt"
	"runtime"
	"sync"
)

func main() {
	var i, tmp, length int
	var a []int

	length = 100
	a = make([]int, length)

	for i = 1; i <= length; i++ {
		a[i-1] = i
	}

	//$omp parallel do
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	chunkSize := length / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= length; start += chunkSize {
		end := start + chunkSize - 1
		if end > length {
			end = length
		}
		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				tmp = a[i-1] + i
				a[i-1] = tmp
			}
		}(start, end)
	}
	wg.Wait()
	//$omp end parallel do

	fmt.Printf("a(50)=%3d\n", a[49])
}
